### PR TITLE
[Continuous batching] In place logit transformations

### PR DESCRIPTION
**Pull request by mzegla**
Changes:
- Changing apply() methods signature to avoid returning vector copies and run all transformations in place
- Change default benchmark app generation config and use dynamic split fuse by default
- Minor fixes

### CODE REVIEW

File: src/cpp/src/logit_processor.hpp [hunk 1]
```diff
@@ -70,20 +66,19 @@ class TemperatureLogitTransform : public ILogitTransformer {
 public:
     TemperatureLogitTransform(double temperature) : m_temperature(temperature) {};
 
-    std::vector<Token> apply(const std::vector<Token>& input_logits) override {
-        std::vector<Token> output(input_logits.begin(), input_logits.end());
-        std::sort(output.begin(), output.end(), [](const Token& lhs, const Token& rhs) {return lhs.m_log_prob > rhs.m_log_prob; });
-        float max_logit = output[0].m_log_prob;
+    void apply(std::vector<Token>& logits) override {
+        // TODO: No need to sort entire vector if we only need max prob
+        std::sort(logits.begin(), logits.end(), [](const Token& lhs, const Token& rhs) {return lhs.m_log_prob > rhs.m_log_prob; });
```
Comment (iefode) [inline src/cpp/src/logit_processor.hpp:71]: Minor: To replace by `max_element`?

File: src/cpp/src/logit_processor.hpp [hunk 2]
```diff
@@ -174,14 +167,14 @@ class EOSPenaltyTransform : public ILogitTransformer {
...
         m_eos_token_id(eos_token_id), m_applicable_tensor_len(min_generated_tokens) {}
 
-    std::vector<Token> apply(const std::vector<Token>& input_logits) {
-        std::vector<Token> output(input_logits.begin(), input_logits.end());
-        for (auto& token_id : output) {
+    void apply(std::vector<Token>& logits) override {
+        for (auto& token_id : logits) {
+            // TODO: Is that right? m_index seems to be only a location of the token that logit is bound to
+            // while here we compare that index to token id
             if (token_id.m_index == m_eos_token_id) {
```
Comment (iefode) [inline src/cpp/src/logit_processor.hpp:174]: Could you please fix it? Looks you are right. Please check `set_eos_token in the generation config`

File: src/cpp/src/logit_processor.hpp [hunk 3]
```diff
@@ -319,19 +306,19 @@ class LogitProcessor {
                 if (sampling_params.top_k > 0) {
                     m_logit_transformers.emplace_back(new LogitTransformers::TopKFilter(sampling_params.top_k));
                 }
+                // TODO: Temperature transorm already does normalization and top_p/top_k do not seem to require it\
+                // It looks obsolete
                 m_logit_transformers.emplace_back(new LogitTransformers::ProbabilityNormalizeTransform());
```
Comment (iefode) [inline src/cpp/src/logit_processor.hpp:311]: Let's check in HF or LLM to avoid of extra code. I have saved this transformer according old sampler code..

**Review by iefode — COMMENTED** [review]
In general, LGTM

File: src/cpp/src/logit_processor.hpp (hunk 1 above)
Comment (mzegla) [reply src/cpp/src/logit_processor.hpp:71]: Sure, I'll do it, but in the next PR I'll likely propose custom loop here due to performance gains I observed in a draft PR.

File: src/cpp/src/logit_processor.hpp (hunk 3 above)
Comment (mzegla) [reply src/cpp/src/logit_processor.hpp:311]: I didn't see explicit normalization done in vLLM sampler and in our code this transformer does not look useful. 
If there's no top_p and top_k then it's just repeating the last part of temperature transformer. If we apply top filters and don't think applying normalization second time on limited set will be helpful. 

Honestly I wonder if we need normalization at all in this case. After temperature is applied the only transforms we have are top_p and top_k that do not modify logits and after normalization the relations between logits don't change (if we sort them we get the same outcome). But perhaps I'm missing something.

File: src/cpp/src/logit_processor.hpp (hunk 2 above)
Comment (mzegla) [reply src/cpp/src/logit_processor.hpp:174]: Actually I think I was wrong here. At some point I've seen that logits size is smaller than vocab size and hence I had a suspicion that this might not work as expected, but I've check that the size is matching so such index to id comparison seems to be valid here.

**Review by iefode — APPROVED** [review]

File: tests/python_tests/test_preemption.py [hunk 4]
```diff
@@ -22,7 +22,6 @@ def get_beam_search_seq_len_300() -> GenerationConfig:
     generation_config.num_beam_groups = 3
     generation_config.num_beams = 6
     generation_config.max_new_tokens = 300
-    generation_config.num_return_sequences = 3
```
Comment (popovaan) [inline tests/python_tests/test_preemption.py:25]: Is this change needed?

Comment (mzegla) [reply tests/python_tests/test_preemption.py:25]: It's just a cleanup. See, that we assign `num_return_sequences` below, so this part was overrode anyway.

**Review by popovaan — APPROVED** [review]

File: samples/cpp/continuous_batching_benchmark/continuous_batching_benchmark.cpp [hunk 5]
```diff
@@ -123,6 +121,11 @@ Dataset filtered_dataset(const std::string& models_path, const std::string& data
 
         ov::genai::GenerationConfig greedy_search = ov::genai::greedy();
         greedy_search.max_new_tokens = std::min(max_output_len, output_len);
+        greedy_search.repetition_penalty = 1.0;
+        greedy_search.frequency_penalty = 0.0;
+        greedy_search.presence_penalty = 0.0;
+        greedy_search.diversity_penalty = 0.0;
+        greedy_search.length_penalty = 0.0;
```
Comment (ilya-lavrenov) [inline samples/cpp/continuous_batching_benchmark/continuous_batching_benchmark.cpp:128]: by default all penalties should be OFF (if it's not a case, we need to fix it), so do we need to set them explicitly here?

Comment (mzegla) [reply samples/cpp/continuous_batching_benchmark/continuous_batching_benchmark.cpp:128]: I suppose they were not OFF at a time this change was made. I agree we should rather change defaults than set it to 0 here.

Comment (ilya-lavrenov) [reply samples/cpp/continuous_batching_benchmark/continuous_batching_benchmark.cpp:128]: Could you please set then to 0 if it's not done yet? 
And update this place accordingly
